### PR TITLE
Fix for issue #376

### PR DESCRIPTION
**Pull request by obrix**
Try fixing the interruption of powsybl process badly quitting the python interpreter by properly detecting the interruption and ending the java process by throwing an error.

Signed-off-by: Bertrand Rix <bertrand.rix@artelys.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

Issue #376 

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Comment by sylvlecl** [comment]
I have been thinking a little more about that issue:

1. Finally, I think we should keep the logback behaviour of "swallowing" exceptions.
This is because a typical use-case of loggers is to be called inside `catch` statements, where we should avoid to rethrow something (in particular in `doCatch` method, otherwise the process crashes again).

2. Then, we should probably still avoid to swallow the keyboard interrupt. In order to achieve this, we could try to restore the python exception at the end of the java call.
Note that this will not actually interrupt the java work, but this is the same behaviour as today without logging.

3. Finally, if we wanted to actually enable the user to interrupt long java calls, it would require a lot more work:
  - translating the python interrupt into java `Thread.interrup()`
  - more importantly, long java functions of powsybl such as loading a network would need to check for interrupts so that it actually does something

**Comment by obrix** [comment]
@sylvlecl Great !

> Finally, I think we should keep the logback behaviour of "swallowing" exceptions.

I think this is still the case in the provided solution

> Then, we should probably still avoid to swallow the keyboard interrupt. In order to achieve this, we could try to restore the python exception at the end of the java call.
Note that this will not actually interrupt the java work, but this is the same behaviour as today without logging.

This cannot be done through a classic exception from AppenderBase.append, this is the reason I used an Error which inherit from Throwable and thus go through AppenderBase.doAppend exception filter. This has the effect of interrupting the long process immediatly. If we want to continue the process and "restore the python exception at the end of the java call" this become more complicated.
Maybe from AppenderBase.append where I throw the error I can instead raise a flag in PypowsyblApiLib that will be used to throw an exception from Util.doCatch after the run. This should implement a behavior similar to what you are describing, not interrupting the process but ending the call with an exception. What do you think ?

**Comment by sylvlecl** [comment]
> I think this is still the case in the provided solution

Not really since we throw the Error! I mean that after all, it's better to not throw at all (or throw a normal exception that logback will catch, but it's not very useful)

> Maybe from AppenderBase.append where I throw the error I can instead raise a flag in PypowsyblApiLib that will be used to throw an exception from Util.doCatch after the run. This should implement a behavior similar to what you are describing, not interrupting the process but ending the call with an exception. What do you think ?

I think to achieve this we could use the python primitives, and bypass completely the java part:

1. if a `py::error_already_set` is thrown by pybind11 in the c++ callback, we can "restore it" to the python C API representation and continue java work normally:
```cpp
catch(py::error_already_set& err) {
    err.restore();
}
```

2. then at the end of the java call, on c++ side, we can check if a python error has been stored:
```cpp
if (PyErr_Occurred() != nullptr) {
    throw py::error_already_set();
}
```
This way the calling python code would receive the inner exception after the java work has completed.

**Comment by obrix** [comment]
@sylvlecl As discussed I tried the described approach in the last commit. Catching the exception and restoring it let the java process go to its end but PyErr_Occured does not allow to rethrow it at the end. This has the effect of disabling interruption with ctrl+c. Do you see anything that differs from what you tested ?

File: cpp/src/pylogging.cpp [hunk 1]
```diff
@@ -35,12 +35,30 @@ py::object CppToPythonLogger::getLogger() {
...
+    ~save_python_error() {
+        if (PyErr_Occurred() == nullptr) {
+            PyErr_Restore(type, value, trace); 
+        }
+    }
+};
+
 void logFromJava(int level, long timestamp, char* loggerName, char* message) {
+    py::gil_scoped_acquire acquire;
+    save_python_error previousError;  // to keep and restore the previously set exception, if any
```
Comment (sylvlecl) [inline cpp/src/pylogging.cpp:53]: @obrix I added this mechanism to allow to keep the initial exception during the subsequent log calls.

It seems that without this, after the log call which sets the error, the previously set error is cleared (the python error is set at the beginning of the call, but not set anymore at the end ...).

With that additional workaround, the KeyboardInterrupt exception is correctly passed all the way until the end of the java call, and re-thrown thanks to your code.
From python side, we correctly receive it:
```
Traceback (most recent call last):
  File "toto.py", line 8, in <module>
    network = pn.load('/home/leclercsyl/feature_tests/01-test-cases/02-full/pf_20200326T1635Z_20200326T1635Z.xiidm')
  File "/home/leclercsyl/Projects/sources/powsybl/pypowsybl/pypowsybl/network.py", line 4058, in load
    return Network(_pp.load_network(file, parameters))
  File "/usr/lib/python3.8/logging/__init__.py", line 1497, in log
    def log(self, level, msg, *args, **kwargs):
KeyboardInterrupt
``` 

It looks good this way :)

**Review by sylvlecl — APPROVED** [review]

File: cpp/src/pylogging.cpp (hunk 1 above)
Comment (obrix) [reply cpp/src/pylogging.cpp:53]: Cool, thank you!

**Pull request merged**